src/02/fp-math.asm

```diff
@@ -0,0 +1,1353 @@
+        push hl \ pop iy
+        ; Check for a negative sign at the beginning
+        ld a, (ix)
+        cp '-'
+        jr nz, _
+        ld (hl), 0x80
+        inc ix
+_:
+        inc hl
+        inc hl
+        ld b, 14    ; Remaining digits counter
+        ld d, 0     ; Parsing flags
+        ld e, 0x80  ; Place value counter
+.loop:
+        ; Check if we are at the end of the buffer
+        xor a
+        cp b
+        jr z, .loopEnd
+        ; Check if we've reached a null terminator
+        ld a, (ix)
+        or a
+        jr z, .loopEnd
+        cp '.'
+        jr nz, _
+        ; Make sure we haven't already seen a decimal point
+        bit 0, d
+        jr nz, .error
+        set 0, d
+        jr .loopCont
+_:
+        ; Error on non-numeric characters
+        cp '9' + 1
+        jr nc, .error
+        sub '0'
+        jr c, .error
+        jr nz, _
+        ; Handle zero
+        bit 1, d
+        jr z, .loopCont     ; Skip leading zero
+_:
+        set 1, d    ; Finished with all leading zeroes
+        rld         ; Shift the digit into the current BCD byte
+        bit 0, b    ; Only increment HL every other digit
+        jr z, .loopIter
+        inc hl
+.loopIter:
+        ; Adjust place value for digits before decimal
+        ld a, d
+        cp 0x02
+        jr nz, _
+        inc e
+_:
+        inc ix
+        djnz .loop
+        jr .loopEnd
+.loopCont:
+        ; Adjust place value for leading zeroes after decimal
+        ld a, d
+        cp 0x01
+        jr nz, _
+        dec e
+_:
+        inc ix
+        jr .loop
+.loopEnd:
+        ; Shift the last digit if there were an odd number
+        bit 0, b
+        jr z, _
+        xor a
+        rld
+_:
+        ; Check if there may be more place values that need counted
+        ld a, b
+        or a
+        jr nz, .remainderLoopEnd
+        bit 0, d
+        jr nz, .remainderLoopEnd
+.remainderLoop:
+        ; Count remaining place values
+        ld a, (ix)
+        or a
+        jr z, .remainderLoopEnd
+        cp '.'
+        jr z, .remainderLoopEnd
+        cp '9' + 1
+        jr nc, .error
+        cp '0'
+        jr c, .error
+        inc e
+        inc ix
+        jr .remainderLoop
+.remainderLoopEnd:
+        ; Correct for one's place value
+        ld a, 0x80
+        cp e
+        jr nc, _
+        dec e
+_:
+        ; Save the place value counter into the exponent byte
+        ld (iy + 1), e
+        pop af
+        cp a    ; Set Z flag
+        jr .exit
+.error:
+        pop af
+        ; Save A
+        ld b, a
+        or 1    ; Reset Z flag
+        ; Restore A
+        ld a, b
+.exit:
+    pop iy
+    pop de
+    pop bc
+    pop hl
+    pop ix
+    ret
+
+;; fptostr [FP Math]
+;;  Converts a floating point number into an ASCII-encoded decimal string.
+;; Inputs:
+;;  IX: Pointer to floating point number
+;;  HL: Pointer to destination buffer
+;;  A: Flags
+;; Notes:
+;;  The destination buffer must be large enough to store the result.
```

Comment:
The worst case I can think of is -1.234567891E-100, which is 17 + terminator = 18 bytes. But for safety I wouldn't make any of my buffers less than 20-25. Out of everything in the file, I probably spent the most time wrangling bugs in `fptostr`.